File: infrastructures/rest-ddd/csharp/src/Application/Errors/ExampleServiceException.cs
/**
 * Example Service Exception
 * Follows ServiceException pattern with error codes.
 */

using System.Collections.Generic;

namespace Application.Errors
{
    public enum ExampleErrorCode
    {
        NotFound,
        ValidationFailed,
        Conflict,
        Forbidden
    }

    public class ExampleServiceException : ServiceException<ExampleErrorCode>
    {
        /**
         * Service exception for Example operations.
         * 
         * Rules:
         * - Located in /Application/Errors/
         * - Pattern: {Object}ServiceException
         * - Define error codes as enum
         * - Define message templates
         */

        private static readonly IReadOnlyDictionary<string, string> MESSAGE_TEMPLATES =
            new Dictionary<string, string>
            {
                { nameof(ExampleErrorCode.NotFound), "Example '{id}' not found" },
                { nameof(ExampleErrorCode.ValidationFailed), "Validation failed: {reason}" },
                { nameof(ExampleErrorCode.Conflict), "Example '{id}' already exists" },
                { nameof(ExampleErrorCode.Forbidden), "Access to Example '{id}' is forbidden" }
            };

        public ExampleServiceException(
            ExampleErrorCode p_errorCode,
            IReadOnlyDictionary<string, object>? p_details = null
        ) : base(p_errorCode, MESSAGE_TEMPLATES, p_details)
        {
        }

        /**
         * Map error code to HTTP status code.
         */
        public static int ToHttpStatus(ExampleErrorCode p_errorCode)
        {
            return p_errorCode switch
            {
                ExampleErrorCode.NotFound => 404,
                ExampleErrorCode.ValidationFailed => 400,
                ExampleErrorCode.Conflict => 409,
                ExampleErrorCode.Forbidden => 403,
                _ => 500
            };
        }
    }
}

File: infrastructures/rest-ddd/typescript/src/application/errors/ExampleServiceException.ts
/**
 * Example Service Exception
 * Follows ServiceException pattern with error codes.
 */

import { ServiceException } from './ServiceException';

export enum ExampleErrorCode {
  NOT_FOUND = 'NOT_FOUND',
  VALIDATION_FAILED = 'VALIDATION_FAILED',
  CONFLICT = 'CONFLICT',
  FORBIDDEN = 'FORBIDDEN'
}

export class ExampleServiceException extends ServiceException<ExampleErrorCode> {
  /**
   * Service exception for Example operations.
   * 
   * Rules:
   * - Located in /application/errors/
   * - Pattern: {Object}ServiceException
   * - Define error codes as enum
   * - Define message templates
   */

  private static readonly MESSAGE_TEMPLATES: Record<ExampleErrorCode, string> = {
    [ExampleErrorCode.NOT_FOUND]: "Example '{id}' not found",
    [ExampleErrorCode.VALIDATION_FAILED]: "Validation failed: {reason}",
    [ExampleErrorCode.CONFLICT]: "Example '{id}' already exists",
    [ExampleErrorCode.FORBIDDEN]: "Access to Example '{id}' is forbidden"
  };

  constructor(
    p_errorCode: ExampleErrorCode,
    p_details?: Record<string, any>
  ) {
    super(p_errorCode, ExampleServiceException.MESSAGE_TEMPLATES, p_details);
  }

  /**
   * Map error code to HTTP status code.
   */
  static toHttpStatus(p_errorCode: ExampleErrorCode): number {
    const mapping: Record<ExampleErrorCode, number> = {
      [ExampleErrorCode.NOT_FOUND]: 404,
      [ExampleErrorCode.VALIDATION_FAILED]: 400,
      [ExampleErrorCode.CONFLICT]: 409,
      [ExampleErrorCode.FORBIDDEN]: 403
    };

    return mapping[p_errorCode] || 500;
  }
}
